hail/src/main/scala/is/hail/expr/ir/DistinctlyKeyed.scala

```diff
@@ -1,71 +1,96 @@
-        memo.bind(node, ())
-        analyze(child, memo)
-      case TableOrderBy(child, _) => analyze(child, memo)
-      case TableRename(child, _, _) => basicChildrenCheck(IndexedSeq(child))
-      case TableFilterIntervals(child, _, _) => basicChildrenCheck(IndexedSeq(child))
-      case TableToTableApply(child, _) => analyze(child, memo)
-      case BlockMatrixToTableApply(_, _, _) =>
-      case BlockMatrixToTable(_) =>
-      case RelationalLetTable(_, _, body) => basicChildrenCheck(IndexedSeq(body))
-      case _: MatrixIR =>
-        throw new IllegalArgumentException("MatrixIR should be lowered when it reaches distinct analysis")
-      case _: BlockMatrixIR =>
-      case ir: IR =>
-        ir.children.foreach(child => analyze(child, memo))
-    }
 
+  def apply(node: BaseIR): DistinctKeyedAnalysis =
+    DistinctKeyedAnalysis(flattenIR(node).foldLeft(Memo.empty[Unit])(analyze))
+
+  /**
+   * Update the memo if the node is distinctly-keyed
+   */
+  private def analyze(memo: Memo[Unit], node: BaseIR): Memo[Unit] = node match {
+    case t: TableRead =>
+      memo.bindIf(t.isDistinctlyKeyed, t, ())
+
+    case t@TableKeyBy(child, keys, _) =>
+      memo.bindIf(child.typ.key.forall(keys.contains) && memo.contains(child), t, ())
+
+    case t@(_: TableFilter | _: TableMapRows | _: TableMapGlobals) =>
+      memo.bindIf(memo.contains(t.children.head), t, ())
+
+    case t@RelationalLetTable(_, _, body) =>
+      memo.bindIf(memo.contains(body), t, ())
+
+    case t@(_: TableHead
+            | _: TableTail
+            | _: TableRepartition
+            | _: TableJoin
+            | _: TableIntervalJoin
+            | _: TableMultiWayZipJoin
+            | _: TableLeftJoinRightDistinct
+            | _: TableRename
+            | _: TableFilterIntervals) =>
+      memo.bindIf(t.children.forall(memo.contains), t, ())
+
+    case t@(_: TableRange
+            | _: TableDistinct
+            | _: TableKeyByAndAggregate
+            | _: TableAggregateByKey) =>
+      memo.bind(t, ())
+
+    case _: MatrixIR =>
+      throw new IllegalArgumentException("MatrixIR should be lowered when it reaches distinct analysis")
+
+    case _ =>
+      memo
   }
 
+  /**
+   * Iterate over the "interesting" nodes in the IR in post-order (ie. only
+   * those nodes that contribute to the analysis of distinctly-keyed nodes).
+   */
+  private def flattenIR: BaseIR => Iterator[BaseIR] =
+    TreeTraversal.postOrder {
+      case t@(_: TableAggregateByKey
+              | _: TableFilter
+              | _: TableKeyByAndAggregate
+              | _: TableMapGlobals
+              | _: TableMapPartitions
+              | _: TableMapRows) =>
+        t.children.take(1).iterator
+
+      case RelationalLetTable(_, _, body) =>
+        Iterator.single(body)
+
+      case t@(_: IR
+              | _: TableDistinct
+              | _: TableExplode
+              | _: TableFilterIntervals
+              | _: TableHead
+              | _: TableIntervalJoin
```

Comment:
you didn't change this behavior, but TableIntervalJoin is actually inherits distinctness just from the left -- it joins intervals from the right onto the left without a cartesian product

Reply:
Thanks! I'll update it